src/core/Akka/Util/Option.cs

```diff
@@ -30,7 +30,7 @@ public struct Option<T>
         public Option(T value)
         {
             Value = value;
-            HasValue = true;
+            HasValue = value != null;
```

Comment:
I was always thinking about that - but what if `null` in a valid value for this optional thing? What if `null` means "there is a value, it is obtained from external source/whatever, it is initialized, and it is `null` right now"?

Reply:
No, by definition Option<T> will return None or Some, what is the point of using Option if you still allow null to leak through? you wouldn't fix the original NRE problem.

Reply:
Even with this, you'll still have NRE problem if you forgot to check for None.

Reply:
In essence, Option is a half assed attempt to fix NRE problem, its just moving the problem somewhere else.

Reply:
If you still need a null, then what you need is a tri-state object, where its state can be null, none, and some

Reply:
And this will mean that `new Option(default(int)).HasValue == true` but `new Option(default(object)).HasValue == false`, which may make `Option` class less obvious... 
I realize the intention to void `new Option(null).HasValue` been `true`, but just thinking - what if this is a feature, not a bug? Besides, this may become a pretty tricky breaking change...

Just need to think twice before merging this change I think.

Reply:
BTW, the scala implementation actually throws a `NoSuchElementException` if you try to access a null value, which i think the right implementation... but changing that would probably break binary compatibility

Reply:
Our implementation is flawed anyway, Option<T> is supposed to be an IOption<T> with Some<T> and None<T> concrete classes

Reply:
Yeah, I agree with all of that, just thinking about how many things we can break because people are passing null (and how often they are doing this by design)

Reply:
No idea, its broken from the start

Reply:
Yeah, it even broke our own code.
This is stupid, I feel like I'm chasing my own tail.

Reply:
This is a good fix actually, I would love to see exception thrown here too - but maybe this should be done on some breaking change release, i.e. Akka.NET 1.5, or even 2.0, because chances are that many users will need to debug their app and fix things.

Reply:
I think the proper thing to do right now is to implement the right `IOption<T>` interface but make it internal and keep the old `Option<T>` struct for public consumption.
But this will require so so many work to convert every `Option<T>` in our code x__x

Reply:
Why do we need the `IOption<T>` interface exactly? I get the `Some<T>` and `None<T>`